nomad/plan_apply_test.go

```diff
@@ -1098,9 +1088,625 @@ func TestPlanApply_EvalNodePlan_Node_Disconnected(t *testing.T) {
+			duplicateAllocIndexErrorString)
+	})
+
+	// Stopping a job produces a destructive plan only, but increments the job
+	// version. We cannot really test for conflicts here, so this can be used
+	// to ensure correct behaviour.
+	t.Run("job stop", func(t *testing.T) {
+		testPlan := structs.Plan{
+			Job: &structs.Job{
+				ID:        testAllocs[0].JobID,
+				Namespace: testAllocs[0].Namespace,
+				Version:   1,
+			},
+			NodeUpdate: map[string][]*structs.Allocation{
+				"8bdb21db-9445-3650-ca9d-0d7883cc8a73": {
+					{Name: testJobGroupID + "[0]"},
+					{Name: testJobGroupID + "[1]"},
+					{Name: testJobGroupID + "[2]"},
+				},
+			},
+			NodeAllocation: nil,
+		}
+		must.NoError(t, evaluatePlanAllocIndexes(testStateSnapshotFn(t, testState), &testPlan))
+	})
+
+	// When a deployment finishes, a noop plan is processed which does not
+	// include any allocation updates.
+	t.Run("deployment finish", func(t *testing.T) {
+		testPlan := structs.Plan{
+			Job: &structs.Job{
+				ID:        testAllocs[0].JobID,
+				Namespace: testAllocs[0].Namespace,
+				Version:   1,
+			},
+			NodeUpdate:     nil,
+			NodeAllocation: nil,
+		}
+		must.NoError(t, evaluatePlanAllocIndexes(testStateSnapshotFn(t, testState), &testPlan))
+	})
+
+	// Canary deployments generate a unique state and plan. This test exercises
+	// the plan generated by a canary deployment being promoted.
+	t.Run("canary promote", func(t *testing.T) {
+
+		// Generate a canary allocation and upset this into state. This mimics
+		// a deployment state which requires manual promotion.
+		canaryAlloc := mock.Alloc()
+		canaryAlloc.Job.Version = 1
+		canaryAlloc.Job = testAllocs[0].Job
+		canaryAlloc.JobID = testAllocs[0].Job.ID
+		canaryAlloc.Name = fmt.Sprintf("%s.%s[%v]", canaryAlloc.JobID, canaryAlloc.TaskGroup, 0)
+
+		must.NoError(t, testState.UpsertAllocs(structs.MsgTypeTestSetup, 20, []*structs.Allocation{canaryAlloc}))
+
+		testPlan := structs.Plan{
+			Job: &structs.Job{
+				ID:        testAllocs[0].JobID,
+				Namespace: testAllocs[0].Namespace,
+				Version:   1,
+			},
+			NodeUpdate: map[string][]*structs.Allocation{
+				"52b3508a-c88e-fc83-74c9-829d5ef1103a": {
+					{Name: testJobGroupID + "[0]", ID: testAllocs[0].ID},
+					{Name: testJobGroupID + "[1]", ID: testAllocs[1].ID},
+				},
+				"8bdb21db-9445-3650-ca9d-0d7883cc8a73": {
+					{Name: testJobGroupID + "[2]", ID: testAllocs[2].ID},
+				},
+			},
+			NodeAllocation: map[string][]*structs.Allocation{
+				"52b3508a-c88e-fc83-74c9-829d5ef1103a": {
+					{Name: testJobGroupID + "[1]"},
+				},
+				"8bdb21db-9445-3650-ca9d-0d7883cc8a73": {
+					{Name: testJobGroupID + "[2]"},
+				},
+			},
+		}
+		must.NoError(t, evaluatePlanAllocIndexes(testStateSnapshotFn(t, testState), &testPlan))
+
+		// Delete the alloc state entry, so it doesn't impact any subsequent
+		// tests.
+		must.NoError(t, testState.DeleteEval(30, []string{}, []string{canaryAlloc.ID}, false))
```

Comment:
Should this (and similar bits in these tests) happen in a `t.Cleanup` so that subsequent tests can run even if a test fails?